src/gmres_e.m

```diff
@@ -0,0 +1,426 @@
+                        dy(:, jj + 1) = abs(imag(dy0(:, ij)) * sqrt(1));
+                        jj = size(dy, 2);
+                        if ij < k
+                            if dy0(:, ij) == conj(dy0(:, ij + 1))
+                                ij = ij + 2;
+                            else
+                                ij = ij + 1;
+                            end
+                        end
+                    end
+                else
+                    dy(:, jj + 1) = dy0(:, ij);
+                    ij = ij + 1;
+                    jj = size(dy, 2);
+                end
+            end
+        else
+            dy = dy0;
+        end
+
+        % Update and restart.
+        restart = restart + 1;
+    end
+
+    % Empty matrix E for next outer iteration
+    E = zeros(s, k);
+
+    % ---> GMRES-E Algorithm for restart > 1 or
+    % GMRES-E(m, k)
+
+    while flag == 0 && restart <= maxit
+
+        % Compute normalized residual vector
+        r = b - A * xm;
+        beta = norm(r);
+        v1 = r / beta;
+
+        % Augmented Gram-Schmidt Arnoldi iteration
+        % Notice that for GMRES-E, we need E,
+        % a n-by-k matrix, the matrix of eigenvectors
+        % from the last outer iteration.
+        [H, V, s] = ...
+            augmented_gram_schmidt_arnoldi(A, v1, m, fliplr(dy(:, 1:k)));
+        % Patch: to be solved
+        % Why does the order matter?
+
+        % Plane rotations
+        [HUpTri, g] = plane_rotations(H, beta);
+
+        % Solve the least-squares problem
+        Rs = HUpTri(1:s, 1:s);
+        gs = g(1:s);
+        minimizer = Rs \ gs;
+
+        % Re-write last k vectors???
+        V(:, m + 1:s) = dy(:, 1:k);
+        x = xm + V * minimizer;
+
+        % Update residual norm, iterations, and relative residual vector
+        res(restart + 1, :) = abs(g(s + 1, 1));
+        iter(restart + 1, :) = restart + 1;
+        relresvec(restart + 1, :) = ...
+            res(restart + 1, :) / res(1, 1);
+
+        % Check convergence
+        if relresvec(restart + 1, 1) < tol
+            % We reached convergence.
+            flag = 1;
+            time = toc();
+            return
+
+        elseif restart < maxit
+
+            % We have not reached convergence.
+            % Eigenvalue problem setup, from [1], p. 1161, step 5
+            W = V(1:n, 1:s);
+            Fold = W' * A' * W;
+            G = Rs' * Rs;
+            opts.tol = tol;
+            opts.v0 = ones(size(Fold, 1), 1);
+            E = zeros(s, k);
+            D = zeros(k, 1);
+
+            % Compute the harmonic Ritz vectors
+            [E2, D2] = eigs(Fold, G, k, 'LM', opts);
+            for p = 1:k
+                D(p, 1) = abs(D2(p, p));
+            end
+            [~, I] = sort(D, 1);
+            for q = 1:k
+                E(:, q) = E2(:, I(q, 1));
+            end
+            dy0 = V * E; % yi = Q * gi???
+
+            dy = [];
+
+            % If dy0 has complex components, we separate each eigenvector
```

Comment:
At first glance, this whole block is identical to the one above. Is there any chance we can encapsulate this into a subfunction?

Reply:
More likely, to be solved in issue #70.